### ISSUE TITLE
Process ads newest to oldest

### ISSUE THREAD

**Issue by willscott** [ipni/storetheindex#764]
currently we download the full ad chain, then go from oldest up to newest.
with some accounting changes we could process ads as we download the ad chain.
that would be more efficient in not needing a temporary ingest buffer.

would need to save state and reconcile multiple 'heads' / which ads have been seen for a provider.

**Comment by masih** [comment]
Note to self: Before start working on this read through this: https://github.com/filecoin-project/storetheindex/pull/224#issue-1143761459

**Comment by gammazero** [comment]
Since the chain has the ability to update the metadata associated with a context ID, only the most recent metadata for a context ID must be used. So during reverse chronological processing (newest to oldest), a map of context ID to metadata must be kept so that only the latest metadata is used.  If processing of an ad chain is stopped and then resumes from where it left off (network error, indexer restart, etc.) then this map must also be recovered or recreated from the chain

**Comment by gammazero** [comment]
If processing of an ad chain is interrupted, then processing must resume at the point in the chain where the indexer left off. This unfortunately cannot be done without first traversing the chain from the newest entry since that is necessary to collect information about deleted context and about the mapping of context ID to metadata (see previous comment).

This also means that indexing could be processing the chain from the newest advertisement (head) down to the newest advertisement already indexed, and from the oldest advertisement already indexed (where previous indexing was interrupted) down to the oldest advertisement in the chain. Indexing on both sides of the chain will add extra complexity.

**Comment by gammazero** [comment]
Processing ads from newest to oldest will only make a difference when it takes a long time to download a long ad chain, and the indexer could be indexing ads before getting the remaining chain. So, this should not matter on a caught-up indexer. Also, even if the indexer is taking a long time to download a chain, that does not mean that the ads can be processed on it right away, because the indexer may be too busy processing ads from other chains anyway.  Since the indexer is usually busy with multiple ad chains, delaying processing ads on one chain does not waste time since there are usually other ads to be processed. 

Processing newest to oldest does add complexity:
- Need to keep map of context ID to metadata
- When resuming, need to re-iterate already processed chain to recreate context ID/metadata map and deleted context ID list.
- Need to keep additional information on incomplete indexing

In conclusion, processing newest to oldest adds complexity, does not matter for caught-up indexers, and does not matter for not-caught-up indexers when the indexer is busy processing ads for multiple publishers. With the changes from PR #1548, there is less need to try to speed up ingestion by processing ads while still downloading the chain. So, I think we should continue to process oldest to newest, and close this issue after including the comments here in a document.

**Comment by gammazero** [comment]
Not doing this for the reasons cited above. Closing.
